### PR TITLE
TEC-16773 Update search by vanityName to return instances according to primaryOrganizationType

### PR DESCRIPTION
**Pull request by kennethwong-hc**
### Description of Changes

- Add primaryOrganizationType in OrganizationBase
- Update findOrganizationByVanityName 
  - return OrganizationBrand if primaryOrganizationType is BRAND
  - return Organization if primaryOrganizationType is NONE or SCHOOL

### Reference
https://docs.microsoft.com/en-us/linkedin/marketing/integrations/community-management/organizations/organization-lookup-api?tabs=http

### Documentation
N/A

### Risks & Impacts
- Add support for organization brands

### Testing
- Modified LinkedInMetaInfo to use new version
```
    OrganizationBase normal = LinkedInMetaInfo.getPageInfoByVanityName(linkedInAPI,
        "Echobox");
    System.out.println(organizationToString(normal));
  
    OrganizationBase showcase = LinkedInMetaInfo.getPageInfoByVanityName(linkedInAPI,
        "bio-basketball-brand");
    System.out.println(organizationToString(showcase));
  
    OrganizationBase school = LinkedInMetaInfo.getPageInfoByVanityName(linkedInAPI,
        "university-of-cambridge");
    System.out.println(organizationToString(school));
  
    OrganizationBase notFound = LinkedInMetaInfo.getPageInfoByVanityName(linkedInAPI,
        "news");
    System.out.println(organizationToString(notFound));
```
Print the type and data
```
  private static String organizationToString(OrganizationBase organization) {
    if (organization == null) {
      return "null";
    }
    StringBuilder sb = new StringBuilder();
    if (organization instanceof Organization) {
      Organization organization1 = (Organization) organization;
      sb.append("[Organization] - ");
      sb.append("vanityName: ").append(organization.getVanityName())
          .append(", primaryOrganizationType: ").append(organization.getPrimaryOrganizationType())
          .append(", locations: ").append(organization1.getLocations())
          .append(", organizationStatus: ").append(organization1.getOrganizationStatus())
          .append(", organizationType: ").append(organization1.getOrganizationType())
          .append(", deleted: ").append(organization1.getDeleted())
          .append(", entityStatus: ").append(organization1.getEntityStatus())
          .append(", groups: ").append(organization1.getGroups())
          .append(", localizedWebsite: ").append(organization1.getLocalizedWebsite())
          .append(", overviewPhotoV2: ").append(organization1.getOverviewPhotoV2())
          .append(", schoolAttributes: ").append(organization1.getSchoolAttributes())
          .append(", specialities: ").append(organization1.getSpecialties())
          .append(", staffCountRange: ").append(organization1.getStaffCountRange())
          .append(", alternativeNames: ").append(organization.getAlternativeNames())
          .append(", coverPhotoV2: ").append(organization.getCoverPhotoV2())
          .append(", defaultLocale: ").append(organization.getDefaultLocale())
          .append(", description: ").append(organization.getDescription())
          .append(", foundedOn: ").append(organization.getFoundedOn())
          .append(", id: ").append(organization.getId())
          .append(", industries: ").append(organization.getIndustries())
          .append(", localizedDescription: ").append(organization.getLocalizedDescription())
          .append(", localizedName: ").append(organization.getLocalizedName())
          .append(", logoV2: ").append(organization.getLogoV2())
          .append(", name: ").append(organization.getName())
          .append(", parentRelationship: ").append(organization.getParentRelationship())
          .append(", URN: ").append(organization.getUrn())
          .append(", versionTag: ").append(organization.getVersionTag())
          .append(", website: ").append(organization.getWebsite());
  
    } else if (organization instanceof OrganizationBrand) {
      OrganizationBrand organizationBrand = (OrganizationBrand) organization;
      sb.append("[OrganizationBrand] - ");
      sb.append("vanityName: ").append(organization.getVanityName())
          .append(", primaryOrganizationType: ").append(organization.getPrimaryOrganizationType())
          .append(", alternativeNames: ").append(organization.getAlternativeNames())
          .append(", coverPhotoV2: ").append(organization.getCoverPhotoV2())
          .append(", defaultLocale: ").append(organization.getDefaultLocale())
          .append(", description: ").append(organization.getDescription())
          .append(", foundedOn: ").append(organization.getFoundedOn())
          .append(", id: ").append(organization.getId())
          .append(", industries: ").append(organization.getIndustries())
          .append(", localizedDescription: ").append(organization.getLocalizedDescription())
          .append(", localizedName: ").append(organization.getLocalizedName())
          .append(", logoV2: ").append(organization.getLogoV2())
          .append(", name: ").append(organization.getName())
          .append(", parentRelationship: ").append(organization.getParentRelationship())
          .append(", URN: ").append(organization.getUrn())
          .append(", versionTag: ").append(organization.getVersionTag())
          .append(", website: ").append(organization.getWebsite())
          .append(", tags: ").append(organizationBrand.getTags())
          .append(", pinnedPost: ").append(organizationBrand.getPinnedPost())
          .append(", autoCreated: ").append(organizationBrand.getAutoCreated());
    }
    return sb.toString();
  }
```
Result:
```
[Organization] - vanityName: echobox, primaryOrganizationType: NONE, locations: [com.echobox.api.linkedin.types.organization.LocationInfo@6869a3b3, com.echobox.api.linkedin.types.organization.LocationInfo@6ab4ba9f, com.echobox.api.linkedin.types.organization.LocationInfo@27ace0b1], organizationStatus: null, organizationType: null, deleted: null, entityStatus: null, groups: null, localizedWebsite: https://www.echobox.com, overviewPhotoV2: null, schoolAttributes: null, specialities: null, staffCountRange: null, alternativeNames: null, coverPhotoV2: null, defaultLocale: null, description: null, foundedOn: null, id: 5268030, industries: null, localizedDescription: null, localizedName: Echobox, logoV2: com.echobox.api.linkedin.types.organization.CroppedImage@664e5dee, name: com.echobox.api.linkedin.types.objectype.MultiLocaleString@431f1eaf, parentRelationship: null, URN: null, versionTag: null, website: null
[OrganizationBrand] - vanityName: bio-basketball-brand, primaryOrganizationType: BRAND, alternativeNames: null, coverPhotoV2: null, defaultLocale: null, description: null, foundedOn: null, id: 80927451, industries: null, localizedDescription: null, localizedName: bio basketball brand, logoV2: com.echobox.api.linkedin.types.organization.CroppedImage@46cb98a3, name: com.echobox.api.linkedin.types.objectype.MultiLocaleString@3ffb3598, parentRelationship: null, URN: null, versionTag: null, website: null, tags: null, pinnedPost: null, autoCreated: null
[Organization] - vanityName: university-of-cambridge, primaryOrganizationType: SCHOOL, locations: [com.echobox.api.linkedin.types.organization.LocationInfo@4a34e9f], organizationStatus: null, organizationType: null, deleted: null, entityStatus: null, groups: null, localizedWebsite: https://www.cam.ac.uk/, overviewPhotoV2: null, schoolAttributes: com.echobox.api.linkedin.types.organization.SchoolAttribute@6f6621e3, specialities: null, staffCountRange: null, alternativeNames: null, coverPhotoV2: null, defaultLocale: null, description: null, foundedOn: null, id: 4522, industries: null, localizedDescription: null, localizedName: University of Cambridge, logoV2: com.echobox.api.linkedin.types.organization.CroppedImage@3fc05ea2, name: com.echobox.api.linkedin.types.objectype.MultiLocaleString@7c891ba7, parentRelationship: null, URN: null, versionTag: null, website: null
null

```

## Final Checklist

Please tick once completed.

- [x] Build passes.
- [x] Versioning considered (the version number in this PR is inline with semantic 
versioning requirements).
- [x] Change log has been updated.

### CODE REVIEW

File: src/main/java/com/echobox/api/linkedin/connection/v2/OrganizationConnection.java [hunk 1]
```diff
@@ -167,8 +170,25 @@ public Organization retrieveOrganization(URN organizationURN, Parameter fields)
...
-    return getListFromQuery(ORGANIZATIONS, Organization.class,
-        parameters.toArray(new Parameter[0]));
+    List<Organization> organizationList = getListFromQuery(ORGANIZATIONS,
+        Organization.class, parameters.toArray(new Parameter[0]));
+    if (organizationList.size() == 0) {
+      return organizationList;
+    }
+    
+    PrimaryOrganizationType targetType = organizationList.get(0).getPrimaryOrganizationType();
+    if (PrimaryOrganizationType.NONE.equals(targetType)) {
```
Comment (djjashton) [inline src/main/java/com/echobox/api/linkedin/connection/v2/OrganizationConnection.java:180]: Should we provide some default logic for SCHOOL OrganizationType?

**Review by djjashton — CHANGES_REQUESTED** [review]
Looks good. I have one small question.

**Review by djjashton — APPROVED** [review]
Looks good

**Comment by kennethwong-hc** [comment]
Over to @eddspencer for final review

**Review by eddspencer — CHANGES_REQUESTED** [review]
Thanks Kenneth just a couple of questions...

File: src/main/java/com/echobox/api/linkedin/connection/v2/OrganizationConnection.java [hunk 2]
```diff
@@ -167,8 +170,28 @@ public Organization retrieveOrganization(URN organizationURN, Parameter fields)
...
     addStartAndCountParams(parameters, null, count);
-    return getListFromQuery(ORGANIZATIONS, Organization.class,
-        parameters.toArray(new Parameter[0]));
+    List<Organization> organizationList = getListFromQuery(ORGANIZATIONS,
+        Organization.class, parameters.toArray(new Parameter[0]));
+    if (organizationList.size() == 0) {
+      return organizationList;
+    }
+    
+    PrimaryOrganizationType targetType = organizationList.get(0).getPrimaryOrganizationType();
```
[the remark this answers is not in the document]
Comment (kennethwong-hc) [reply src/main/java/com/echobox/api/linkedin/connection/v2/OrganizationConnection.java]: Created a new class containing all fields from `Organization `and `OrganizationBrand `for deserialising. The instances in the result list will be created according to their `primaryOrganizationType`.
This also saves one endpoint call for organization brands.

File: src/main/java/com/echobox/api/linkedin/types/organization/OrganizationResult.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      organization.setDeleted(this.getDeleted());
+      organization.setEntityStatus(this.getEntityStatus());
+      organization.setGroups(this.getGroups());
+      organization.setLocalizedWebsite(this.getLocalizedWebsite());
+      organization.setLocations(this.getLocations());
+      organization.setSpecialties(this.getSpecialties());
+      organization.setStaffCountRange(this.getStaffCountRange());
+      organization.setSchoolAttributes(this.getSchoolAttributes());
+      organization.setOverviewPhotoV2(this.getOverviewPhotoV2());
+      organization.setOrganizationType(this.getOrganizationType());
```
Comment (eddspencer) [inline src/main/java/com/echobox/api/linkedin/types/organization/OrganizationResult.java:81]: What about `organizationStatus`?

**Review by eddspencer — CHANGES_REQUESTED** [review]
Thanks Kenneth - I think we may have one field missing....I was wondering whether the non-V2 fields should be included as well but I see that the [LinkedIn API docs](https://docs.microsoft.com/en-us/linkedin/marketing/integrations/community-management/organizations/organization-lookup-api?tabs=http#organization) do not include these so probably fine to leave them out

File: src/main/java/com/echobox/api/linkedin/types/organization/OrganizationResult.java (hunk 3 above)
Comment (kennethwong-hc) [reply src/main/java/com/echobox/api/linkedin/types/organization/OrganizationResult.java:83]: Good spot! Thanks.

**Comment by kennethwong-hc** [comment]
Hi @eddspencer , yah, I think it is better to keep propagating those non-V2 fields, as it is an open-source SDK. It should be no harm if LinkedIn doesn't return those fields.
Besides, I added a unit test for the conversion. However, it is required to expose setUrn(). Do you know why we don't expose this setter? (this is the only field that doesn't have setter). The only reason I could think of is URN should be unchangeable, but id should be unchangeable as well ...

**Review by eddspencer — APPROVED** [review]
That looks great thanks Kenneth - no I was not part of the group that made the decision to not have a URN setter. However, I suspect you are right.....but by that logic you do not want to really change most of the fields so happy to have the URN setter existing.

**Pull request merged**